File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
     int n=nums.size();
        k%=n;//for when k>n   
        reverse(nums.begin(), nums.begin()+n-k);//from[1,2,3,4,5,6,7]-> [4,3,2,1,5,6,7]
        reverse(nums.begin()+n-k, nums.end());//from [4,3,2,1,5,6,7] -> [4,3,2,1,7,6,5]
        reverse(nums.begin(), nums.end());//from  [4,3,2,1,7,6,5] -> [5,6,7,1,2,3,4]
    }
        
};

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        length = len(nums)
        k = k%length
        end = (length - k) - 1
        i = 0
        while i < end:
            nums[i] , nums[end] = nums[end], nums[i]
            i += 1
            end -= 1
        
        i = length - k
        end = length - 1
        while i < end:
            nums[i] , nums[end] = nums[end], nums[i]
            i += 1
            end -= 1
        
        end = length - 1
        i = 0
        while i < end:
            nums[i] , nums[end] = nums[end], nums[i]
            i += 1
            end -= 1
        
        